### PR TITLE
input: add some initial keyboard matrix library stubs

### PR DESCRIPTION
**Pull request by fabiobaltieri**
Picking up the work from https://github.com/zephyrproject-rtos/zephyr/issues/52355 and https://github.com/zephyrproject-rtos/zephyr/pull/42923, making this an internal library (similar to `i2c_bitbang` or `can_mcan`) rather than a subsystem api.

---

We currently have three keyboard scanning drivers in the code base (npcx, it8xxx2 and mchp_xec, last two yet to be converted to input). These have been largely copy pasted from each other and share a lot of the same structure and code.

This PR lays a foundation to start decoupling feature from those drivers into a common code base, and it is heavily inspired by the current regulator common data/config one.

Feature wise this only moves the thread struct, stack and initialization to the common code and declares the thread callback as the only API, but the intent is to move as much code as possible in there an only abstract device specific callbacks in the api structures.

### CODE REVIEW

File: drivers/input/input_kbd_matrix.h [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+/**
+ * @brief Common keyboard matrix data.
+ *
+ * This structure **must** be placed first in the driver's data structure.
+ */
+struct input_kbd_matrix_common_data {
+	struct k_thread thread;
+
+	K_KERNEL_STACK_MEMBER(thread_stack, CONFIG_INPUT_KBD_THREAD_STACK_SIZE);
```
Comment (fabiobaltieri) [inline drivers/input/input_kbd_matrix.h:54]: I finally understand why everyone puts a space before this one, checkpatch complains. Should really fix checkpatch, but Perl.

File: drivers/input/Kconfig.kbd_matrix [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+# SPDX-License-Identifier: Apache-2.0
+
+config INPUT_KBD_MATRIX
+	bool
+	help
+	  Enable library used for keyboard matrix drivers.
+
+if INPUT_KBD_MATRIX
+
+config INPUT_KBD_THREAD_STACK_SIZE
```
Comment (keith-zephyr) [inline drivers/input/Kconfig.kbd_matrix:11]: Nit - 
```suggestion
config INPUT_KBD_MATRIX_THREAD_STACK_SIZE
```

Comment (fabiobaltieri) [reply drivers/input/Kconfig.kbd_matrix:11]: Fixed, thanks for catching this.

**Review by ChiHuaL — DISMISSED** [review]

**Review by MulinChao — DISMISSED** [review]

**Review by keith-zephyr — DISMISSED** [review]

**Comment by fabiobaltieri** [comment]
Merge conflict, rebased.

**Review by keith-zephyr — APPROVED** [review]

**Review by MulinChao — APPROVED** [review]

**Pull request merged**